File: leetcode/C++/stone-game-vii.cpp
class Solution {
public:
    int traverse(vector<vector<int>> &dp ,vector<int> &s,int sum, int l, int r){
        if(l>=r) return 0;
        else if(dp[l][r]) return dp[l][r];
        return dp[l][r]=max(sum-s[l] - traverse(dp, s, sum-s[l], l+1, r),sum-s[r] - traverse(dp, s,sum-s[r], l, r-1));
    }
    int stoneGameVII(vector<int>& stones) {
        int s=0;
        vector<vector<int>> dp(stones.size(),vector<int>(stones.size(),0));
        for(int &t : stones) s+=t;
        return traverse(dp,stones,s,0,stones.size()-1);
    }
};

File: leetcode/Python/stone-game-vii.py
class Solution:
    def stoneGameVII(self, stones: List[int]) -> int:
        dp = [[0 for i in range(len(stones))] for j in range(len(stones))]
        def _traverse(l,r,s):
            if l>=r:
                return 0
            if dp[l][r]>0:
                return dp[l][r]
            dp[l][r]=max(s-stones[l]-_traverse(l+1,r,s-stones[l]), s-stones[r]-_traverse(l,r-1,s-stones[r]))
            return dp[l][r]
        return _traverse(0,len(stones)-1,sum(stones))